src/NuGet.Clients/NuGet.SolutionRestoreManager/RestoreManagerPackage.cs

```diff
@@ -114,5 +137,50 @@ private bool ShouldRestoreOnBuild
                 return packageRestoreConsent.IsAutomatic;
             }
         }
+
+        #region IVsUpdateSolutionEvents4
+
+        public void UpdateSolution_QueryDelayFirstUpdateAction(out int pfDelay)
+        {
+            // check if NuGet lock is already acquired by some other NuGet operation
+            if (LockService.IsLockHeld)
+            {
+                // delay build by setting pfDelay to non-zero
+                pfDelay = 1;
+            }
+            else
+            {
+                var isLockAcquired = LockService.TryAcquireLock(NuGetLockTimeout, out _nuGetLock);
+
+                if (isLockAcquired)
+                {
+                    // Set delay to 0 which means don't delay build
+                    pfDelay = 0;
+                }
+                else
+                {
+                    // delay build since NuGet lock is still held by other operation
+                    pfDelay = 1;
+                }
+            }
+        }
+
+        public void UpdateSolution_BeginFirstUpdateAction() { }
+
+        public void UpdateSolution_EndLastUpdateAction()
+        {
+            // Release NuGet lock so that other NuGet operation can continue now.
+            _nuGetLock?.Dispose();
```

Comment:
I won't get into the debate, except to say that I agree you should probably be as defensive as possible with regard to when you expect these events to be raised. There are so many kinds of builds (batch build, canceled builds, skipped builds, builds that are denied the right to even start, etc.) that each raise events in different ways, skipping event handlers, etc... And since you have basically no time to give this any road testing for escrow, I think you should be as defensive as possible.